---
title: "UBC iGEM 2022 - Differential Expression Analysis"
output:
  html_document:
    df_print: paged
---
### Differential Expression Workflow:
1. Read in the Data
2. Create a DGEList Object
3. Filter
4. Normalize
5. Estimate Dispersions
6. Differential Expression

```{r, results='hide', message=FALSE, warning=FALSE}
library(tidyverse)
library(edgeR)
library(limma)
library(dplyr)
library(goseq)

x <- read.delim("data/SRP045409_tpm.tsv.gz", row.names="gene")
x <- x %>% rename(HD_6h_1=SRR1542417, HD_6h_2=SRR1542416, HD_1h_1=SRR1542415, HD_1h_2=SRR1542414, H_6h_1=SRR1542413, H_6h_2=SRR1542412, H_1h_1=SRR1542411, H_1h_2=SRR1542410, D_6h_1=SRR1542409, D_6h_2=SRR1542408, D_1h_1=SRR1542407, D_1h_2=SRR1542406, ctrl_1=SRR1542405, ctrl_2=SRR1542404)

```

```{r}
# round counts to nearest integer
rounded_data <- round(x)

d <- DGEList(counts=rounded_data, 
             group=factor(
               c("HD_6h", "HD_6h", "HD_1h", "HD_1h", "H_6h", "H_6h", "H_1h", "H_1h", "D_6h", "D_6h", "D_1h", "D_1h", "ctrl", "ctrl")))
d
```
```{r}
keep <- filterByExpr(d)
d <- d[keep, , keep.lib.sizes=FALSE]
d <- calcNormFactors(d)
d
```
```{r, results='hide', message=FALSE, warning=FALSE}
d <- estimateDisp(d)
```

As there are only a few samples, it's difficult to estimate the dispersion accurately for each gene, so we need a way of "sharing" information between genes. Possible solutions include:

- Using a common estimate across all genes (common dispersion).

- Fitting an estimate based on the mean-variance trend across the dataset, such that genes w/ similar abundances have similar variance estimates (trended dispersion).

- Computing a genewise dispersion (tagwise dispersion)

```{r}
plotBCV(d)
```

```{r}
et <- exactTest(d, pair=c("HD_6h","ctrl"))
# just looking at the top 100 hits for now
HD6h_vs_ctrl <- topTags(et, n=100)
save(HD6h_vs_ctrl, file='HD6h_ctrl_100.RData')
```
```{r}
de <- decideTestsDGE(et, adjust.method="BH", p.value=0.001)
summary(de)
```

```{r}
detags <- rownames(d)[as.logical(de)] 
plotSmear(et, de.tags=detags)
abline(h = c(-2, 2), col = "blue")
```

```{r}
hist(et$table[,"PValue"], breaks=50, xlab="P-value", main="P-value Distribution")
```
```{r}
sigResults <- et$table[et$table$PValue<0.001,]
dim(sigResults)
```
```{r}
# up-regulated and down-regulated gene expression in HD_6h group compared to ctrl
sigReg <- HD6h_vs_ctrl$table[HD6h_vs_ctrl$table$FDR<0.01,]
sigUpReg <- filter(sigReg, logFC < -1)
sigDownReg <- filter(sigReg, logFC > 1)
sigUpReg <- sigUpReg[order(sigUpReg$logFC),]
sigDownReg <- sigDownReg[order(sigDownReg$logFC, decreasing=TRUE), ]
print(sigUpReg)
print(sigDownReg)
```

```{r}
eff.lib.size <- d$samples$lib.size*d$samples$norm.factors
normCounts <- cpm(d)
pseudoNormCounts <- log2(normCounts + 1)
boxplot(pseudoNormCounts, las=3)
```
```{r}
# plot based on multi-dimensional scaling
plotMDS(pseudoNormCounts)
```
```{r}
# HD6h vs. ctrl
volcanoData <- cbind(et$table$logFC, -log10(et$table$PValue))
colnames(volcanoData) <- c("log(FC)", "-log10(Pvalue)")
plot(volcanoData, pch=20)
abline(v=c(-2, 2), col="blue", lwd=3, lty=2)
abline(h=-log10(0.001), col="red", lwd=3, lty=2)
write.csv(et$table,"HD6H_ctrl.csv", row.names = TRUE)
```
### Gene Ontology (GO) and KEGG Pathway Enrichment Analysis

```{r}
# keg <- kegga(et, species.KEGG="taes")
# topKEGG(keg, sort="up")
# topKEGG(keg, sort="down")
```

